rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -215,50 +467,310 @@ mod tests {
+        let test_runtime = TestRuntime::new();
+        let exporter = ExporterWrapper::with_not_send(
+            ExporterWithNotSendEffectHandler::without_send_test(test_runtime.counters()),
+            test_runtime.config(),
+        );
+
+        test_runtime
+            .set_exporter(exporter)
+            .run_test(scenario())
+            .run_validation(validation_procedure());
+    }
+
+    #[test]
+    fn test_exporter_with_send_effect_handler() {
+        let test_runtime = TestRuntime::new();
+        let exporter = ExporterWrapper::with_send(
+            ExporterWithSendEffectHandler::with_send_test(
+                test_runtime.counters(),
+                |effect_handler| {
+                    exec_in_send_env(|| {
+                        _ = effect_handler.exporter_name();
+                    });
+                },
+            ),
+            test_runtime.config(),
+        );
+
+        test_runtime
+            .set_exporter(exporter)
+            .run_test(scenario())
+            .run_validation(validation_procedure());
+    }
+
+    fn make_chan() -> (
+        mpsc::Sender<ControlMsg>,
+        mpsc::Sender<String>,
+        MessageChannel<String>,
+    ) {
+        let (control_tx, control_rx) = mpsc::Channel::<ControlMsg>::new(10);
+        let (pdata_tx, pdata_rx) = mpsc::Channel::<String>::new(10);
+        (
+            control_tx,
+            pdata_tx,
+            MessageChannel::new(control_rx, pdata_rx),
+        )
+    }
+
+    #[tokio::test]
+    async fn test_control_priority() {
+        let (control_tx, pdata_tx, mut channel) = make_chan();
+
+        control_tx
+            .send_async(ControlMsg::Ack { id: 1 })
+            .await
+            .unwrap();
+        pdata_tx.send_async("pdata1".to_owned()).await.unwrap();
+
+        // Control message should be received first due to bias
+        let msg = channel.recv().await.unwrap();
+        assert!(matches!(msg, Message::Control(ControlMsg::Ack { id: 1 })));
+
+        // Then pdata message
+        let msg = channel.recv().await.unwrap();
+        assert!(matches!(msg, Message::PData(ref s) if s == "pdata1"));
+    }
+
+    #[tokio::test]
+    async fn test_shutdown_drain() {
+        let (control_tx, pdata_tx, mut channel) = make_chan();
+
+        // Pre-load pdata
+        pdata_tx.send_async("pdata1".to_string()).await.unwrap();
+        pdata_tx.send_async("pdata2".to_string()).await.unwrap();
+
+        // Send shutdown with a deadline
+        control_tx
+            .send_async(ControlMsg::Shutdown {
+                deadline: Duration::from_millis(100), // 100ms deadline
+                reason: "Test Shutdown".to_string(),
+            })
+            .await
+            .unwrap();
+
+        // Send more pdata after shutdown is sent, but before receiver likely gets it
+        pdata_tx.send_async("pdata3".to_string()).await.unwrap();
+        sleep(Duration::from_millis(10)).await; // Give receiver a chance to see shutdown
```

Comment:
I guess this sleep is not required. Since we are calling `await` on all the send methods, we know that the control message for Shutdown has made it to the channel, which means as soon as the `recv` method is called, we would first receive the shutdown message.
```suggestion
```

Reply:
Nice